Updates to SparkFun Ublox Arduino Library

averywallis opened the pull request:
Hello,
First time contributing to a public repository, so please let me know if I do anything wrong/suggestions for better way to convey this information in the future.

This pull request is to provide some new functionality to the library. Over the summer, I interned for a company that used the Sparkfun GPS-RTK2 board (ZED-F9P) in a GPS system. Because of the large board size, I designed a board to incorporate the GPS and LoRa radios to create a base-rover GPS system. The company did not want to use u-center to configure the GPS every time or have to update configurations upon a possible hardware failure. For this reason, we used the SparkFun Ublox library to configure the ZED-F9P. Unfortunately the library did not have all the functionality we needed. I spent 2 months understanding the Ublox protocol, this Arduino library and its format, the ZED-F9P, and u-center. I used this information to create the additional GPS functions that our company required for this system.

The following are major functions I have added. Note that these were ONLY developed for the ZED-F9P, so I have no way to confirm if they work on other Ublox GPS modules.
- Set GPS into Fixed mode with a passed longitude, latitude, altitude, and accuracy
- Enable or disable a specific NMEA message for a specific port, while leaving previous ports untouched
- Enable or disable a specific UBX message out a specific port
- Set the Nav5 Dynamic Platform Model
- Set the NMEA Talker ID
- Renamed a large number of Macro definitions in the .h file. Not all are used, but may be useful for future versions
- Added some structs to store assorted GPS data

Note that I just forked the current version (10/15/19) of this library, and then replaced the .cpp and .h files with the edited versions I made. The version I edited was downloaded in early June, 2019, so there will differences of both my edits and changes that have happened on my end. Hopefully this isn't too much of a problem and you are able to see what I have added.

I hope some of this is usable. If I had the GPS module, I would be happy to continue working on this library in my free time. Unfortunately, I had to leave all materials (minus this code) at the company when I had to return to school, so I have been unable to do any additional work with this software.
No idea how others give me feedback about my work, so I will be trying check GitHub on a regular basis.

nseidle commented:
Hi Avery - Welcome! And thank you for this PR. I really appreciate it.

First, let me say, I *welcome* the help to add features and improvements. Let's discuss them in an issue perhaps.

Feature: Enable or disable a specific NMEA message for a specific port, while leaving previous ports untouched - Yes please!
Feature: Enable or disable a specific UBX message out a specific port - It's been awhile but I believe we already have this functionality.
Feature: Added some structs to store assorted GPS data - Awesome. Do a separate PR.
Feature: Set GPS into Fixed mode with a passed longitude, latitude, altitude, and accuracy - Great! New PR.

Any new feature or thing you do in a PR, please write/provide an example in the PR that demos that feature. 

Just a few points that I would need addressed before accepting this PR:

* This PR is very large. I cannot tell what you changed or intend to change. Please consider closing this PR then make many PRs with small, unique changes. 
* Don't change defines such as [this](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files?file-filters%5B%5D=.h#diff-9fdceffd9a523941a1d33c38658e9d63L121). You're breaking reverse compatibility. Imagine a user who uses that define in their code. If they update this lib to your version, it will break their code.
* Adding comments like [this](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files?file-filters%5B%5D=.h#diff-9fdceffd9a523941a1d33c38658e9d63R138) is superb and much appreciated. Consider a separate PR for these.
* Why did you remove [debugPrint](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files#diff-09b68aa0742e5325ae8fb037d77970ebL94)?
* Don't remove [my comments](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files?file-filters%5B%5D=.h#diff-9fdceffd9a523941a1d33c38658e9d63L157) that help me remember my previous mistakes ;)
* Please don't change the way we [define bit fields](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files?file-filters%5B%5D=.h#diff-9fdceffd9a523941a1d33c38658e9d63L159). We do this to be very sure the correct bit is in the correct location. Your method is valid but more error prone. 
* Why are you removing a [large number of methods](https://github.com/sparkfun/SparkFun_Ublox_Arduino_Library/pull/43/files?file-filters%5B%5D=.h#diff-9fdceffd9a523941a1d33c38658e9d63L264). This will break code for many users. Perhaps I'm viewing the PR changes incorrectly.
* This is a nit pick, but your PR title doesn't tell me what you changed. Be more descriptive like "add XYZ feature" or "move debug statements to blarg"

averywallis commented:
Hey Nathan,
Thanks for the feedback! I see a lot of the things you have problems with and I agree with your points. I think some of the problems that you pointed out are due to the large differences in version between what I tried to merge and what is currently on the repository. Looking at the file history there were about 20 commits between when I first downloaded the library at work and when I created this PR. My guess is a bunch of functionality and code was changed between those commits, causing some stuff to appear as if I deleted it because I never actually had it in my code to begin with because it's an older version. I will spend some time this weekend looking through the current version of code in this repo and figuring out where my code fits in.

I think I should mention that the features I added were built only using the packet structure, with the class, ID, length, starting spot. These methods are technically deprecated with more recent protocol versions, as stated in the [interface description](https://www.u-blox.com/sites/default/files/u-blox_ZED-F9P_InterfaceDescription_%28UBX-18010854%29.pdf) document for the ZED-F9P. I should technically use the setval/getval fucntions, as those are the ones still used, but none of my code uses the setval/getval method. I never got around to understanding that method. That may be something that I used the older packet method because of the [u-center](https://www.u-blox.com/en/product/u-center) packet console and binary console where you can see the exact contents being sent/received. This helped me a lot in creating some new functions.
Just an FYI, I have only tested my code/modifications with the [ZED-F9P GPS module](https://www.u-blox.com/en/product/zed-f9p-module), and only over UART1 to a [SAMD21G18 MCU](https://www.microchip.com/wwwproducts/en/ATsamd21g18). I have not tested it with other u-blox GPS modules, other protocols, or other controllers.

Also, what is your process for verifying code that I have written? I would be happy to try and provide other resources that show that what I built works. But like I said originally, I no longer have any of the hardware to test/verify new code on, so I'm not entirely sure how I would go about that.

I will probably close this PR at some point in the next couple of days so that you have a chance to see what I've written here. Like I said before, I'm new to using GitHub beyond a place to store code, so let me know if there's something I should or shouldn't do or other stuff I should look into figuring out how to use. This weekend I will make some smaller PRs like you suggested, assuming I can figure what caused all the problems.

nseidle commented:
Cool. Be sure to checkout how to rebase a branch:

![image](https://user-images.githubusercontent.com/117102/69280268-3da4b280-0ba3-11ea-97ce-9b8fd8e75483.png)

That should fix most of the discrepancies.

I also used the old ublox method rather than the new getVal/setVal in most of my examples and library code. I'd prefer PRs with getVal/setVal but don't be afraid to use the old method. Added functionality is better than future proofness.

Don't worry about testing across multiple ublox modules. I can test against a few (SAM-M8Q and NOE) and ublox is pretty good about functions working on different platforms as long as we follow the spec.

Verifying code: I'll run your PR against a variety of modules (as mentioned). Please write a MVP example sketch for each new feature. I'll run your sketches then file them into the example structure so that other's can see how to use your methods.

Github: You're doing great! Feel free to close/open whatever. You can do no wrong on this repo.

nseidle commented:
Hi Avery - I'm going to close your PR because the core has changed quite a bit. Please consider contributing in the future! We really enjoy the help.

averywallis commented:
Sorry I didn't get back to this sooner, I was busy with school finals and then holiday stuff with family. I hope to post some new PRs soon.

nseidle commented:
No worries. I know the feeling!
